### PR TITLE
#2226 - fix the group-chat heisenbug

### PR DESCRIPTION
**Pull request by SebinSong**
work on #2226

### CODE REVIEW

**Review by Silver-IT — CHANGES_REQUESTED** [review]
Nice work, @SebinSong.

But I don't think this PR fixes the real heisenbug. Even though, the test passed three times in a row, I can't see any updates which fixes the reason of this heisenbug.

The main reason is because INTERACTIVE message is sometimes created after the NOTIFICATION message. Of course, I think it doesn't happen while we are testing, but only happens in Cypress. So the order of two messages (INTERACTIVE, NOTIFICATION) is random and Cypress can not pick which element to assert with.

The previous code of `cy.wait(1000)` was to wait until two messages are all created. But if we remove this code, the last message could be something that we don't know. Neither the NOTIFICATION, nor the INTERACTIVE message. Because the cypress test code could be reached before those two messages are created.

**Comment by SebinSong** [comment]
@Silver-IT 

>  I can't see any updates which fixes the reason of this heisenbug.

I actually did man. It kinda feels like you left a comment straight away without reading the code..

> The main reason is because INTERACTIVE message is created after the NOTIFICATION message. Of course, I think it doesn't happen while we are testing, but only happens in Cypress. So the order of two messages (INTERACTIVE, NOTIFICATION) is random and Cypress can not pick which element to assert with.

I read your inline comment in the code detailing what causes the heisenbug so made below updates based on that, which seems like fixing the bug.

```js
cy.get(messageSelectors.last).invoke('attr', 'class').then(classNames => {
    const isLastMsgTypeNotification = classNames.includes('is-type-notification')
    assertKickerAndMessageContent(
      isLastMsgTypeNotification ? messageSelectors.last : messageSelectors.secondLast
    )
  })
```

**Comment by Silver-IT** [comment]
> The previous code of cy.wait(1000) was to wait until two messages are all created. But if we remove this code, the last message could be something that we don't know. Neither the NOTIFICATION, nor the INTERACTIVE message. Because the cypress test code could be reached before those two messages are created.

@SebinSong, what about this then?

**Comment by SebinSong** [comment]
@Silver-IT I removed that `cy.wait(1000)` cuz it didn't seem to be needed.

**Comment by Silver-IT** [comment]
I can see your updates are same as the previous one. With only one difference which is that you created and used a new function `assertKickerAndMessageContent`.

**Comment by SebinSong** [comment]
@Silver-IT
I called that assertion within the `then()` block is the difference. (when the check for whether it's the last one or the second last one is needed)

**Comment by SebinSong** [comment]
@Silver-IT I thought you did a good job of checking if it's the last message or the second last one via the `classNames`. My updates are only enhancement on top of that (along with removing `cy.wait()`). So, calm down bro..

**Comment by Silver-IT** [comment]
Okay, I will try to test some more with your changes in [this branch](https://github.com/okTurtles/group-income/tree/2226-heisenbug-in-group-chatspecjs-more-persistent-one).

File: test/cypress/integration/group-chat.spec.js [hunk 1]
```diff
@@ -51,26 +51,29 @@ describe('Group Chat Basic Features (Create & Join & Leave & Close)', () => {
...
-      let isLastElement = true
       if (byProposal) {
         // NOTE: when the member is kicked from the from by proposal
         //       two messages will be created in general chatroom; INTERACTIVE, and NOTIFICATION
         //       INTERACTIVE message should be created before the NOTIFICATION message
         //       but sometimes (only in Cypress) NOTIFICATION message could be created earlier
-        //       this block is to handle that heisenbug
-        cy.wait(1000) // eslint-disable-line cypress/no-unnecessary-waiting
-        cy.get('div.c-message:last-child').invoke('attr', 'class').then(classNames => {
-          isLastElement = classNames.includes('is-type-notification')
```
Comment (SebinSong) [inline test/cypress/integration/group-chat.spec.js:64]: It looked to me that the intention here was, some actions needed to happen based on the check result of this `isLastElement`. but the fact that the check is happening within the `then()` block whereas the "some actions" are placed outside that was something looking strange and I felt creating the bug. so, made this updates.

(Leaving this PR for other tasks anyways)

**Comment by SebinSong** [comment]
@Silver-IT btw, let's keep `assertKickerAndMessageContent` function I created. I think it simplifies the logic.

**Comment by Silver-IT** [comment]
@SebinSong Yeah, I agree on keeping that function which simplifies the code nicely. 👍

**Comment by Silver-IT** [comment]
I think this PR reduces the percentage for the heisenbug to happen. (But it still happens actually. [here](https://cloud.cypress.io/projects/q6whky/runs/2750/test-results?utm_source=github&statuses=%5B%7B%22value%22%3A%22FAILED%22%2C%22label%22%3A%22FAILED%22%7D%5D))

> Why it reduces the percentage?

I think it's because it asserts inside the `then` clause so that the two `last messages` (one to invoke class names, and one to assert with) are same, even though they are not in real. This also makes sense only when the `NOTIFICATION` message is created before the `INTERACTIVE` message and the test case is reached after the `NOTIFICATION` message is created.

> Then, when it still happens?

When we pick the last message and invoke class of it **before all the messages are loaded**. This was the reason why I used `cy.wait(1000)`.  Shortly, `isLastMsgTypeNotification` is not always right value.

**Comment by SebinSong** [comment]
Seems like there is another PR created for the same issue. Closed.

**Pull request closed without merging**